### ISSUE TITLE
Make failed: Undefined reference to Gotcha functions and UnifyFS Gotcha-related symbols.

### ISSUE THREAD

**Issue by HenryYihengXu** [LLNL/UnifyFS#643]
### System information

Type                   | Version/Name
---                    |     ---
Operating System       | Linux
OS Version             |  RHEL
Architecture           | Lassen IBM Power9
UnifyFS Version        | Dev
Compiler | gcc@8.3.1

### Describe the problem you're observing
I built UnifyFS dependencies via Spack and built UnifyFS itself with Autotools from the GitHub dev branch. I was following the instruction in the doc: https://unifyfs.readthedocs.io/en/latest/build.html#build-dependencies-with-spack-build-unifyfs-with-autotools.

However, in this way, it wouldn't define `UNIFYFS_GOTCHA` so Gotcha-related codes were not compiled. So I added `-DUNIFYFS_GOTCHA` and gotcha lib and include paths to `CPPFLAGS` and `LDFLAGS`. This time it compiled Gotcha-related codes, but make failed due to undefined reference to Gotcha lib functions and many UnifyFS Gotcha-related symbols, such as setup_gotcha-wrappers, wrappee_handle_fopen, etc.
 
### Describe how to reproduce the problem

I'm using `gcc@8.3.1` on Lassen.

1. use spack to install and load all dependencies
```
. spack/share/spack/setup-env.sh
spack load gotcha%gcc@8.3.1
spack load argobots%gcc@8.3.1
spack load mercury%gcc@8.3.1
spack load mochi-margo%gcc@8.3.1
spack load spath%gcc@8.3.1
```
2. ./autogen.sh
3. Export environment variables:
```
export CPPFLAGS="-I/path/to/gotcha/include -DUNIFYFS_GOTCHA -DHAVE_GOTCHA"
export CXXFLAGS="-I/path/to/gotcha/include -DUNIFYFS_GOTCHA -DHAVE_GOTCHA"
export CFLAGS="-I/path/to/gotcha/include -DUNIFYFS_GOTCHA -DHAVE_GOTCHA"
export LDFLAGS=-L/path/to/gotcha/lib64
``` 
4. ./configure
```
./configure --prefix=/path/to/unifyfs/install \
--with-gotcha=/path/to/gotcha/install \
--with-spath=/path/to/spath/install
```
5. make
### Include any warning or errors or relevant debugging data
During make, I saw the following errors:

```
...
...
...
/usr/tce/packages/gcc/gcc-8.3.1/rh/usr/bin/../libexec/gcc/ppc64le-redhat-linux/8/ld: warning: libjson-c.so.5, needed by /g/g92/xu23/apps/spack/opt/spack/linux-rhel7-power9le/gcc-8.3.1/mochi-margo-0.9.4-2px5oeg5gj6ggi24vw3oyjzdsshopzoq/lib/libmargo.so, may conflict with libjson-c.so.2
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-sysio.o): warning: the `getwd' function is dangerous and should not be used.
/usr/tce/packages/gcc/gcc-8.3.1/rh/usr/bin/../libexec/gcc/ppc64le-redhat-linux/8/ld: warning: libjson-c.so.5, needed by /g/g92/xu23/apps/spack/opt/spack/linux-rhel7-power9le/gcc-8.3.1/mochi-margo-0.9.4-2px5oeg5gj6ggi24vw3oyjzdsshopzoq/lib/libmargo.so, may conflict with libjson-c.so.2
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs.o): In function `unifyfs_init':
unifyfs.c:(.text+0x50b8): undefined reference to `setup_gotcha_wrappers'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o): In function `__wrap_fopen':
unifyfs-stdio.c:(.text+0x1ffc): undefined reference to `gotcha_get_wrappee'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o): In function `__wrap_freopen':
unifyfs-stdio.c:(.text+0x2160): undefined reference to `gotcha_get_wrappee'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o): In function `__wrap_setvbuf':
unifyfs-stdio.c:(.text+0x22d8): undefined reference to `gotcha_get_wrappee'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o): In function `__wrap_setbuf':
unifyfs-stdio.c:(.text+0x242c): undefined reference to `gotcha_get_wrappee'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o): In function `__wrap_ungetc':
unifyfs-stdio.c:(.text+0x274c): undefined reference to `gotcha_get_wrappee'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):unifyfs-stdio.c:(.text+0x28b0): more undefined references to `gotcha_get_wrappee' follow
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x10): undefined reference to `wrappee_handle_fopen'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x20): undefined reference to `wrappee_handle_freopen'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x30): undefined reference to `wrappee_handle_setvbuf'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x40): undefined reference to `wrappee_handle_setbuf'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x50): undefined reference to `wrappee_handle_ungetc'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x60): undefined reference to `wrappee_handle_fgetc'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x70): undefined reference to `wrappee_handle_fputc'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x80): undefined reference to `wrappee_handle_getc'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x90): undefined reference to `wrappee_handle_putc'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0xa0): undefined reference to `wrappee_handle_fgets'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0xb0): undefined reference to `wrappee_handle_fputs'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0xc0): undefined reference to `wrappee_handle_fread'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0xd0): undefined reference to `wrappee_handle_fwrite'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0xe0): undefined reference to `wrappee_handle_vfprintf'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0xf0): undefined reference to `wrappee_handle_vfscanf'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x100): undefined reference to `wrappee_handle_fseek'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x110): undefined reference to `wrappee_handle_fseeko'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x120): undefined reference to `wrappee_handle_ftell'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x130): undefined reference to `wrappee_handle_ftello'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x140): undefined reference to `wrappee_handle_rewind'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x150): undefined reference to `wrappee_handle_fgetpos'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x160): undefined reference to `wrappee_handle_fsetpos'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x170): undefined reference to `wrappee_handle_fflush'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x180): undefined reference to `wrappee_handle_feof'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x190): undefined reference to `wrappee_handle_ferror'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x1a0): undefined reference to `wrappee_handle_clearerr'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x1b0): undefined reference to `wrappee_handle_fileno'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x1c0): undefined reference to `wrappee_handle_fclose'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x1d0): undefined reference to `wrappee_handle_vfwprintf'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x1e0): undefined reference to `wrappee_handle_vfwscanf'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x1f0): undefined reference to `wrappee_handle_fgetwc'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x200): undefined reference to `wrappee_handle_fgetws'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x210): undefined reference to `wrappee_handle_fputwc'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x220): undefined reference to `wrappee_handle_fputws'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x230): undefined reference to `wrappee_handle_fwide'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x240): undefined reference to `wrappee_handle_getwc'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x250): undefined reference to `wrappee_handle_putwc'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-stdio.o):(.toc+0x260): undefined reference to `wrappee_handle_ungetwc'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-sysio.o): In function `__wrap_access':
unifyfs-sysio.c:(.text+0x8a4): undefined reference to `gotcha_get_wrappee'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-sysio.o): In function `__wrap_mkdir':
unifyfs-sysio.c:(.text+0xb10): undefined reference to `gotcha_get_wrappee'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-sysio.o): In function `__wrap_rmdir':
unifyfs-sysio.c:(.text+0xd74): undefined reference to `gotcha_get_wrappee'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-sysio.o): In function `__wrap_chdir':
unifyfs-sysio.c:(.text+0xed4): undefined reference to `gotcha_get_wrappee'
unifyfs-sysio.c:(.text+0xfb4): undefined reference to `gotcha_get_wrappee'
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-sysio.o):unifyfs-sysio.c:(.text+0x155c): more undefined references to `gotcha_get_wrappee' follow
../../client/src/.libs/libunifyfs.a(libunifyfs_la-unifyfs-sysio.o):(.toc+0x10): undefined reference to `wrappee_handle_access'
...
...
...
```

### Other Attempts
I also tried the following things. None of them worked and I got the same errors.
1. Compiled the gotcha to both static and shared libs, because I saw many errors are from building a static libunifyfs.a.
2. Added `--enable-stati=-no`, same reason.
3. Modified `./m4/gotcha.m4` to be:
```
AC_DEFUN([UNIFYFS_AC_GOTCHA], [
  # preserve state of flags
  GOTCHA_OLD_CFLAGS=$CFLAGS
  GOTCHA_OLD_CXXFLAGS=$CXXFLAGS
  GOTCHA_OLD_LDFLAGS=$LDFLAGS

  AC_ARG_WITH([gotcha], [AC_HELP_STRING([--with-gotcha=PATH],
    [path to installed libgotcha [default=/usr/local]])],
    [
      GOTCHA_DIR="${withval}"
      GOTCHA_CFLAGS="-I${GOTCHA_DIR}/include -DHAVE_GOTCHA"
      GOTCHA_LDFLAGS="-L${GOTCHA_DIR}/lib64 -L${GOTCHA_DIR}/lib -Wl,-rpath,${GOTCHA_DIR}/lib64 -Wl,-rpath,${GOTCHA_DIR}/lib -lgotcha"
      CFLAGS="$CFLAGS ${GOTCHA_CFLAGS}"
      CXXFLAGS="$CXXFLAGS ${GOTCHA_CFLAGS}"
      LDFLAGS="$LDFLAGS ${GOTCHA_LDFLAGS}"
    ],
    [
      GOTCHA_CFLAGS=""
      GOTCHA_LDFLAGS=""
    ]
  )

  AC_CHECK_LIB([gotcha], [gotcha_wrap],
    [
      GOTCHA_LIBS="${GOTCHA_LDFLAGS} -lgotcha"
      AC_SUBST(GOTCHA_CFLAGS)
      AC_SUBST(GOTCHA_LDFLAGS)
      AC_SUBST(GOTCHA_LIBS)
      AM_CONDITIONAL([HAVE_GOTCHA], [true])
    ],
    [
      AC_MSG_WARN([couldn't find a suitable libgotcha, use --with-gotcha=PATH])
      AM_CONDITIONAL([HAVE_GOTCHA], [false])
    ],
    []
  )

  # restore flags
  CFLAGS=$GOTCHA_OLD_CFLAGS
  CXXFLAGS=$GOTCHA_OLD_CXXFLAGS
  LDFLAGS=$GOTCHA_OLD_LDFLAGS
])
```
Basically I added `-DHAVE_GOTCHA` and `-lgotcha`, in case gotcha libs cannot be found.

Overall, it looks like there are some problems in your makefiles that it cannot find gotcha libs or even Gotcha-related functions written in your code.

**Comment by HenryYihengXu** [comment]
Thanks to Hari and Cameron's helps. It turns out to be not all parts of UnifyFS are compiled with this flag. The libunifyfs_gotcha.la uses Gotcha but libunifyfs.la doesn't. I applied this flag globally so those who don't need Gotcha would give errors. Also, the assumption that UnifyFS is not using Gotcha is wrong. I checked that in the wrong place (as some code won't use gotcha)
